source/val/validate_graph.cpp

```diff
@@ -0,0 +1,605 @@
+           << output_index << " but graph only has " << graph_type_num_outputs
+           << " outputs.";
+  }
+
+  uint32_t graph_type_output_type =
+      GraphTypeInstGetOutputAtIndex(graph_type_inst, output_index);
+
+  if (has_element_index) {
+    // Check ElementIndex is allowed
+    if (!IsTensorArray(_, graph_type_output_type)) {
+      return _.diag(SPV_ERROR_INVALID_DATA, inst)
+             << "OpGraphSetOutputARM ElementIndex not allowed when the graph "
+                "output selected by "
+             << "OutputIndex is not an OpTypeArray or OpTypeRuntimeArray";
+    }
+
+    // Check ElementIndex is in range if it can be evaluated and the output is a
+    // fixed-sized array whose Length can be evaluated
+    uint64_t element_index;
+    if (_.IsArrayType(graph_type_output_type) &&
+        _.EvalConstantValUint64(inst->GetOperandAs<uint32_t>(2),
+                                &element_index)) {
+      uint64_t array_length;
+      auto graph_type_output_type_inst = _.FindDef(graph_type_output_type);
+      if (_.EvalConstantValUint64(graph_type_output_type_inst->word(3),
+                                  &array_length)) {
+        if (element_index >= array_length) {
+          return _.diag(SPV_ERROR_INVALID_DATA, inst)
+                 << "OpGraphSetOutputARM ElementIndex out of range. The type "
+                    "of the graph output being accessed "
+                 << _.getIdName(graph_type_output_type) << " is an array of "
+                 << array_length << " elements but " << "ElementIndex is "
+                 << element_index;
+        }
+      }
+    }
+  }
+
+  // Check Value's type matches with graph type
+  uint32_t value = inst->word(1);
+  uint32_t value_type = _.FindDef(value)->type_id();
+  if (has_element_index) {
+    uint32_t expected_type = _.GetComponentType(graph_type_output_type);
+    if (value_type != expected_type) {
+      return _.diag(SPV_ERROR_INVALID_DATA, inst)
+             << "The type " << _.getIdName(value_type)
+             << " of Value provided to the graph output instruction "
+             << _.getIdName(value) << " does not match the component type "
+             << _.getIdName(expected_type) << " of output " << output_index
+             << " in the graph type.";
+    }
+  } else {
+    if (value_type != graph_type_output_type) {
+      return _.diag(SPV_ERROR_INVALID_DATA, inst)
+             << "The type " << _.getIdName(value_type)
+             << " of Value provided to the graph output instruction "
+             << _.getIdName(value) << " does not match the type "
+             << _.getIdName(graph_type_output_type) << " of output "
+             << output_index << " in the graph type.";
+    }
+  }
+  return SPV_SUCCESS;
+}
+
+spv_result_t ValidateGraphEnd(ValidationState_t& _, const Instruction* inst) {
+  size_t end_inst_num = inst->LineNum() - 1;
+
+  // Check the graph contains a single output instruction just before the
```

Comment:
This would be easy to add to the combined check to ensure at least one output is used.